packages/rpc-transport/src/params-patcher.ts

```diff
@@ -0,0 +1,26 @@
+type Patched<T> = T extends object ? { [Property in keyof T]: Patched<T[Property]> } : T extends bigint ? number : T;
+
+// FIXME(https://github.com/microsoft/TypeScript/issues/33014)
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type TypescriptBug33014 = any;
+
+export function patchParamsForSolanaLabsRpc<T>(params: T): Patched<T> {
+    if (Array.isArray(params)) {
+        return params.map(patchParamsForSolanaLabsRpc) as TypescriptBug33014;
+    } else if (typeof params === 'object' && params !== null) {
+        const out = {} as TypescriptBug33014;
+        for (const propName in params) {
+            if (Object.prototype.hasOwnProperty.call(params, propName)) {
+                out[propName] = patchParamsForSolanaLabsRpc(params[propName]);
+            }
+        }
+        return out as TypescriptBug33014;
+    } else if (typeof params === 'bigint') {
+        // TODO(solana-labs/solana/issues/30341) Create a data type to represent u64 in the Solana
+        // JSON RPC implementation so that we can throw away this entire patcher instead of unsafely
+        // downcasting `bigints` to `numbers`.
+        return Number(params) as TypescriptBug33014;
```

Comment:
If it's outside of the number range, how about throwing an error?

Reply:
Oooh. Interesting. Yeah, there are three ways to handle numbers above `Number.MAX_SAFE_INTEGER`.

1. Let the cast happen and pass it on.
2. Warn on the console, cast, and pass it on.
3. Throw.

The only options that are API compatible with the current `@solana/web3.js` are the first two. The third would be a breaking change.

There's a fourth, more arduous option: creating an opaque datatype for safe numbers.

```ts
type SafeNumber = bigint & {__safenumber: unique symbol}

function assertIsSafeNumber(num: bigint): asserts num is SafeNumber {
  if (num > Number.MAX_SAFE_INTEGER) {
    throw new Error('...');
  }
}
function asSafeNumber(num: bigint): SafeNumber {
  assertIsSafeNumber(num);
  return num as SafeNumber;
}
```

…and then everywhere you ever wanted to use a number with the RPC…

```ts
rpc.getBlock(asSafeNumber(slot));
```

I think the best solution is the one that requires people to do _nothing_ when we eventually fix the RPC, other than to ask people to update `web3.js`.

Thoughts?

Reply:
If this is meant to be a new library that people consciously switch to, or a new major version of web3.js, then I think it's fine to break and go the route of number 3.  As it stands, people will get garbage anyway, right?

On the flipside, I see what you mean. it would really stink to get an error because you happen to query an account with a lot of lamports, for example. Maybe number 2 isn't so bad then?

Number 4 does seem quite painful, and I think I'm missing something -- how is it much different from number 3 in the end?

Reply:
Number 3 and 4 both kill your program at runtime, but number 4 pushes the assertion out of the framework and into userspace.

This _is_ meant to be a new library, but I _would_ like to be able to rebuild the existing `web3.js` on top of the modern core in the meantime. If the RPC started to throw here, there's nothing that legacy `web3.js` could do to say ‘no no, forget it, just send the cast'd value through.’

Ha. I could literally have the transport take in a handler that's like `onIntegerOverflow`. The legacy one could be a  no-op, and the modern one could throw.

🤔

Reply:
> number 4 pushes the assertion out of the framework and into userspace.

Ah I see, so you're thinking that by forcing people to make a code change like `asSafeNumber` in front of everything, people will think twice? I unfortunately think it'll just cause churn

> Ha. I could literally have the transport take in a handler that's like onIntegerOverflow. The legacy one could be a no-op, and the modern one could throw.

Oooh I really like this!